### ISSUE TITLE
Possible to retrieve git information during build?

### ISSUE THREAD

**Issue by jodyhagins** [Carthage/Carthage#895]
I have a script that runs as part of a framework build, which dynamically creates a .m file containing version information.  It includes information obtained from

    branchName=$(git rev-parse --abbrev-ref HEAD)
    buildNumber=$(git rev-list HEAD --count)
    gitSHA=$(git rev-parse HEAD)
    versionString=$(git describe --tags --always --dirty)

This allows the information to be accurate relative to the exact build.  It works great (adds some custom fields to the plist and generates source code to be linked with the framework).

However, when using the framework via carthage, the repository gets "checked out", but it is done in such a manner where there is no longer a git repository in the checkout.  Thus, those commands either fail or give git information for the git repository in the top-level project itself.

I can see where the "gitness" is stripped in CarthageKit/Git.swift, and I don't see any hooks in there to allow callouts to script that can do some work before the git metadata is blown away.

Could this be added?  Or something that will save "informational" metadata in the carthage subdirectory somewhere so that it can be obtained?

Or, is there some other way around this so that this information can be obtained at build time?

Thanks.

**Comment by mdiep** [comment]
Have you tried using the `--use-submodules` flag to do this?

**Comment by jodyhagins** [comment]
Yes, it works with submodules, but I don't want to force framework users to use submodules.

**Comment by mdiep** [comment]
Unfortunately, I don't think this is a good fit for Carthage. While Carthage does support branch builds, they should only really be used as a last resort since they don't provide version information. For proper releases, frameworks can include the most relevant information inside the repository.

**Comment by jodyhagins** [comment]
I understand your concern, Matt, and I appreciate your response.  However, branch building is only supplemental, and is really a red-herring for this discussion.

Carthage uses the git tag (and SHA) as version information.  I want to make sure my framework automatically uses the git tag as version information as well.  This information is only known after a git commit has taken place.  Once you change the files to include that information, it requires another git commit, which changes the information.

One of the main goals of carthage is to use existing tools and provide a bridge of sorts between git and xcode, "a dependency manager that gets the job done without taking over the responsibility of Xcode, and without creating extra work for framework authors."

Having to manually maintain information that is already in git is a mistake waiting to happen (and one that has already happened unfortunately).  A person shouldn't have to manually duplicate information, especially when the information is readily available for a script.

Being able to access the git metadata can ease the work of framework authors, reduce the probability of a mistake on the part of the framework author, and does not interfere with xcode.

Furthermore, having some way to gain access to git metadata maintains consistency of being able to run automated build scripts between submodule-mode and non-submodule mode.

Thanks!

**Comment by jodyhagins** [comment]
I agree that my request is out of scope for Carthage.

So, let me take a different plan of attack.  Is there a definitive specification for Cartfile.resolved somewhere.  I looked at the code, and can see what it assumes, but is there something documented that can be easily tracked for change?

I wrote a script that "recognizes" it is being built with Carthage, and then parses the Cartfile.resolved to tell it where to go and ask for the metadata.

This seems to work fine, but I'd feel more comfortable knowing there was a definitive spec for Cartfile.resolved somewhere.
